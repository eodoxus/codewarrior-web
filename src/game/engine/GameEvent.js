import Event from "../../lib/Event";

const listeners = [];
let inputQueue;

export default class GameEvent extends Event {
  static CANCEL = "cancel";
  static CLICK = "click";
  static CLICK_HERO = "clickHero";
  static CLICK_MENU = "clickMenu";
  static CLOSE_CURTAIN = "closeCurtain";
  static CLOSE_HERO_MENU = "closeHeroMenu";
  static CLOSE_TATTERED_PAGE = "closeTatteredPage";
  static CONFIRM = "confirm";
  static COLLISION = "collision";
  static DIALOG = "dialog";
  static DOORWAY = "doorway";
  static HIDE_BORDER = "hideBorder";
  static NPC_INTERACTION = "npcInteraction";
  static TALK = "talk";
  static TRANSITION = "transition";
  static OPEN_CURTAIN = "openCurtain";
  static OPEN_HERO_MENU = "openHeroMenu";
  static OPEN_TATTERED_PAGE = "openTatteredPage";
  static SHOW_BORDER = "showBorder";
  static STOP = "stop";

  static click(tile) {
    return new EventType(GameEvent.CLICK, tile);
  }

  static generic(eventName, data) {
    return new EventType(eventName, data);
  }

  static collision(entity) {
    return new EventType(GameEvent.COLLISION, entity);
  }

  static heroClick(hero) {
    return new EventType(GameEvent.CLICK_HERO, hero);
  }

  static menuClick(position) {
    return new EventType(GameEvent.CLICK_MENU, position);
  }

  static talk(entity) {
    return new EventType(GameEvent.TALK, entity);
  }

  static removeAllListeners() {
    listeners.forEach(listener => {
      listener.remove();
    });
  }

  static on(event, handler, enqueue = false) {
    const listener = super.on(event, handler);
    if (enqueue) {
      listeners.push(listener);
    }
    return listener;
  }

  static inputQueue(event) {
    if (!inputQueue) {
      inputQueue = new InputQueue();
    }
    return inputQueue;
  }
}

class EventType {
  type;
  data;
  constructor(type, data) {
    this.type = type;
    this.data = data;
  }

  getType() {
    return this.type;
  }

  getData() {
    return this.data;
  }
}

class InputQueue {
  queue;

  constructor() {
    this.queue = [];
  }

  add(event) {
    if (this.queue.length === 0) {
      this.queue.push(event);
      return;
    }
    if (this.queue[this.queue.length - 1].getType() !== event.getType()) {
      this.queue.push(event);
    }
  }

  getNext() {
    if (this.queue.length) {
      return this.queue.pop();
    }
  }
}
