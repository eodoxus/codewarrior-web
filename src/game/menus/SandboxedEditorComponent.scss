.container {
  position: absolute;
  width: 100%;
  height: 100%;

  iframe {
    width: 100%;
    height: 100%;
    border: none;
  }

  .executeButton {
    position: absolute;
    bottom: -40px;
    left: 50%;
    margin-left: -70px;
  }
}
