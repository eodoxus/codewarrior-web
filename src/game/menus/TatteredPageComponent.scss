@import "../../lib/lib.scss";

$spell-icon-size: 20px;

.container {
  width: 400px;
  height: 329px;
  position: absolute;
  top: 10px;
  left: 120px;
  box-shadow: 0 0 20px #000;
  cursor: text;
  z-index: 3;
  padding: 5px;
  display: flex;

  .buttons {
    position: absolute;
    right: 20px;
    display: flex;
    z-index: 3;

    .spellButton {
      position: relative;
      cursor: pointer;
      color: $color-font-inverse;
      font-size: 10px;
      line-height: 1.8;
      text-align: center;
      background: url("/sprites/hero-menu.png") no-repeat -20px 0;
      width: $spell-icon-size;
      height: $spell-icon-size;
      margin: 5px 5px 0 0;
    }

    .spellButtonOverlay {
      $color-overlay: rgba(94, 92, 92, 0.4);

      position: absolute;
      top: 0;
      left: 0;
      background-color: $color-overlay;
      border: 2px solid transparent;
      border-radius: 15px;
      width: $spell-icon-size;
      height: $spell-icon-size;
    }

    .activeSpell,
    .spellButtonOverlay:hover {
      $border-offset: -2px;
      $icon-size-with-border: $spell-icon-size + 4;

      background: transparent;
      top: $border-offset;
      left: $border-offset;
      border-color: $color-font-inverse;
      width: $icon-size-with-border;
      height: $icon-size-with-border;
    }
  }

  .close {
    position: absolute;
    top: -15px;
    right: -15px;
    cursor: pointer;
    padding: 10px;
    z-index: 1;

    span {
      display: block;
      width: 20px;
      height: 20px;
      border: 2px solid #fff;
      border-radius: 10px;
      background: #000;
      color: $color-font-inverse;
      text-align: center;
      line-height: 1;
    }
  }

  .sandbox {
    position: absolute;
    width: 98%;
    height: 100%;
  }

  button {
    background-color: #e7d9b2;
    border-color: #873c35;
    box-shadow: 0 0 20px #000;
  }

  button:hover {
    background-color: #f3e6c2;
    border-color: #873c35;
  }
}
