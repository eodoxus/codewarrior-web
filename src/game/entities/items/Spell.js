import GameEvent from "../../engine/GameEvent";
import GameScriptModel from "../../../data/GameScriptModel";
import GameState from "../../GameState";
import Interpreter from "js-interpreter";
import Audio from "../../engine/Audio";

const MAX_EXECUTION_STEPS = 9999;
const MAX_MEMORY = 1000000;
const MAX_STACK_SIZE = 99;
const SANITY_CHECK_INTERVAL = 100;
const OUT_OF_MAGIC = "Out of magic";

export default class Spell {
  static _log = "";

  static getLog() {
    return Spell._log;
  }

  static log(...args) {
    if (Spell._log) {
      Spell._log += "\n";
    }
    Spell._log += args.join(" ");
  }

  static setLog(log) {
    Spell._log = log;
  }

  script;
  restults;

  constructor(code = "") {
    this.script = new GameScriptModel({
      data: code
    });
  }

  getApi() {
    return GameState.getSceneApi();
  }

  getCode() {
    return this.script.data;
  }

  setCode(code) {
    this.script.data = code;
  }

  getCost() {
    return this.cost;
  }

  getScript() {
    return this.script;
  }

  setScript(script) {
    this.script = script;
  }

  isMockMode() {
    return GameState.getHero()
      .getBehavior()
      .isReading();
  }

  async cast() {
    Spell.setLog("");
    this.cost = 0;
    try {
      const interpreter = this.createInterpreter(this.getApi());
      await executeCode(interpreter);
      if (this.isMockMode()) {
        return GameEvent.fireAfterClick(GameEvent.DIALOG, Spell.getLog());
      }
      GameEvent.fire(GameEvent.SPELL_CAST, this);
    } catch (e) {
      if (e.message === OUT_OF_MAGIC) {
        return;
      }
      Spell.log(e.message);
      GameEvent.fireAfterClick(GameEvent.DIALOG, {
        error: true,
        msg: Spell.getLog()
      });
      throw new Error("Spell cast failure");
    }
  }

  createInterpreter(api) {
    const apiSandbox = this.sandboxApi(api);
    return new Interpreter(this.getCode(), apiSandbox);
  }

  sandboxApi(api) {
    return (interpreter, scope) => {
      Object.keys(api).forEach(objName => {
        const obj = api[objName];
        const apiObj = interpreter.createObject(Interpreter.OBJECT);
        interpreter.setProperty(scope, objName, apiObj);
        obj.getFunctions().forEach(functionName => {
          if (functionName[0] === "~") {
            functionName = functionName.substr(1);
            this.wrapAsyncFn(interpreter, apiObj, obj, functionName);
          } else {
            this.wrapFn(interpreter, apiObj, obj, functionName);
          }
        });
      });
      this.addLogFn(interpreter, scope);
    };
  }

  wrapFn(interpreter, scope, obj, functionName) {
    const wrapper = (...pseudoArgs) => {
      const nativeArgs = pseudoToNativeArgs(interpreter, pseudoArgs);
      this.accumulateMagicCost(obj, functionName);
      return interpreter.nativeToPseudo(obj[functionName](...nativeArgs));
    };
    interpreter.setProperty(
      scope,
      functionName,
      interpreter.createNativeFunction(wrapper),
      Interpreter.NONENUMERABLE_DESCRIPTOR
    );
  }

  wrapAsyncFn(interpreter, scope, obj, functionName) {
    const wrapper = (...pseudoArgs) => {
      const callback = pseudoArgs.pop();
      const nativeArgs = pseudoToNativeArgs(interpreter, pseudoArgs);
      nativeArgs.push((...args) => {
        callback(interpreter.nativeToPseudo(...args));
      });
      this.accumulateMagicCost(obj, functionName);
      obj[functionName](...nativeArgs);
    };
    interpreter.setProperty(
      scope,
      functionName,
      interpreter.createAsyncFunction(wrapper),
      Interpreter.NONENUMERABLE_DESCRIPTOR
    );
  }

  accumulateMagicCost(obj, functionName) {
    if (this.isMockMode()) {
      return;
    }
    this.cost += obj.getCost(functionName);
    if (GameState.getHero().magic < this.cost) {
      Audio.play(Audio.EFFECTS.OUT_OF_MAGIC);
      throw new Error(OUT_OF_MAGIC);
    }
  }

  addLogFn(interpreter, scope) {
    interpreter.setProperty(
      scope,
      "log",
      interpreter.createNativeFunction((...pseudoArgs) => {
        let nativeArgs = [];
        for (var i = 0; i < pseudoArgs.length; i++) {
          nativeArgs[i] = JSON.stringify(
            interpreter.pseudoToNative(pseudoArgs[i])
          );
        }
        console.log(nativeArgs);
        this.log(nativeArgs);
      })
    );
  }

  edit() {
    GameEvent.fire(GameEvent.OPEN_TATTERED_PAGE, this);
    this.onDoneEditing(this.save);
  }

  onDoneEditing(callback) {
    let listener = GameEvent.on(GameEvent.CLOSE_TATTERED_PAGE, () => {
      callback();
      listener.remove();
      listener = null;
    });
  }

  save = async () => {
    const gameSave = await GameState.getGameSave();
    this.script.game_save_id = gameSave.id;
    return this.script.save();
  };
}

function executeCode(interpreter) {
  return new Promise((resolve, reject) => {
    let numSteps = 0;
    (function step() {
      try {
        executionSanityChecks(interpreter, numSteps++);
        if (interpreter.step()) {
          setTimeout(step);
        } else {
          resolve();
        }
      } catch (e) {
        reject(e);
      }
    })();
  });
}

function executionSanityChecks(interpreter, numSteps) {
  if (numSteps % SANITY_CHECK_INTERVAL === 0) {
    limitMemory(interpreter);
    limitRunTime(numSteps);
    limitStackSize(interpreter);
  }
}

function limitMemory(interpreter) {
  const json = serialize(interpreter);
  if (json.length > MAX_MEMORY) {
    throw new Error("The incantation has grown too large, halting");
  }
}

function limitRunTime(numSteps) {
  if (numSteps > MAX_EXECUTION_STEPS) {
    throw new Error("It's taking too long, incantation will halt");
  }
}

function limitStackSize(interpreter) {
  const stackSize = interpreter.stateStack.length;
  if (stackSize > MAX_STACK_SIZE) {
    throw new Error("Infinite recursion detected, halting");
  }
}

function pseudoToNativeArgs(interpreter, pseudoArgs) {
  const nativeArgs = [];
  for (var i = 0; i < pseudoArgs.length; i++) {
    nativeArgs[i] = interpreter.pseudoToNative(pseudoArgs[i]);
  }
  return nativeArgs;
}

function serialize(interpreter) {
  let cache = [];
  const serialized = JSON.stringify(interpreter, function(key, value) {
    if (typeof value === "object" && value !== null) {
      if (cache.indexOf(value) !== -1) {
        // Circular reference found, discard key
        return;
      }
      // Store value in our collection
      cache.push(value);
    }
    return value;
  });
  cache = null;
  return serialized;
}
