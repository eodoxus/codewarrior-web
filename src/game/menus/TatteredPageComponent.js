import React from "react";
import cx from "classnames";
import styles from "./TatteredPageComponent.scss";
import GameEvent from "../engine/GameEvent";
import GameState from "../GameState";
import MenuComponent from "./MenuComponent";
import Url from "../../lib/Url";
import TextureCache from "../engine/TextureCache";
import SandboxedEditor from "./SandboxedEditorComponent";
import Audio from "../engine/Audio";

export default class TatteredPageComponent extends MenuComponent {
  constructor(props) {
    super(props);
    this.state = {
      image: `${Url.PUBLIC}/images/tattered-page.png`
    };
  }

  componentDidMount() {
    this.listeners = [
      GameEvent.on(GameEvent.OPEN_TATTERED_PAGE, this.onOpen),
      GameEvent.on(GameEvent.CLOSE_TATTERED_PAGE, this.onClose)
    ];
    TextureCache.fetch(this.state.image);
  }

  onClose = e => {
    if (!this.isOpen()) {
      return;
    }
    GameState.setIsReading(false);
    Audio.play(Audio.EFFECTS.CLOSE_BOOK);
    GameEvent.fire(GameEvent.CLOSE_DIALOG);
    this.spell.setCode(this.editor.getCode());
    this.setState({ isOpen: false });
  };

  onOpen = spell => {
    this.spell = spell;
    if (this.isOpen()) {
      return;
    }
    GameState.setIsReading(true);
    Audio.play(Audio.EFFECTS.OPEN_BOOK);
    GameEvent.fire(GameEvent.CLOSE_HERO_MENU);
    this.setState({
      isOpen: true,
      spell
    });
  };

  onSpellClick = spell => {
    this.setState({ spell });
    const code = spell.getCode();
    this.editor.setCode(code);
  };

  render() {
    if (!this.state.isOpen) {
      return "";
    }

    return (
      <div
        className={cx(styles.container, "tattered-page")}
        style={{
          backgroundImage: `url(${this.state.image})`
        }}
        onClick={GameEvent.absorbClick}
      >
        <div
          className={styles.close + " close"}
          onClick={() => GameEvent.fire(GameEvent.CLOSE_TATTERED_PAGE)}
        >
          <span>x</span>
        </div>
        <div className={styles.buttons}>{this.renderButtons()}</div>
        <div className={styles.sandbox}>
          <SandboxedEditor
            code={this.state.spell.getCode()}
            api={this.state.spell.getApi()}
            ref={editor => (this.editor = editor)}
          />
        </div>
      </div>
    );
  }

  renderButtons() {
    const tatteredPage = GameState.getHero().getSpellEditor();
    return tatteredPage.getSpells().map((spell, iDx) => {
      const isActive = this.state.spell.getScript().id === spell.getScript().id;
      return (
        <div
          className={styles.spellButton}
          onClick={() => this.onSpellClick(spell)}
          key={iDx}
        >
          <span>{iDx + 1}</span>
          <div
            className={cx(
              styles.spellButtonOverlay,
              isActive ? styles.activeSpell : ""
            )}
          />
        </div>
      );
    });
  }
}
