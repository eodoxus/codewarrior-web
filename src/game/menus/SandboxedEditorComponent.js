import React, { Component } from "react";
import styles from "./SandboxedEditorComponent.scss";
import cx from "classnames";
import CodeMirror from "react-codemirror";
import GameEvent from "../engine/GameEvent";
import { Button } from "../../components/forms/controls/buttons/index";
import Indicators from "../../components/indicators";
import SandboxedEditorHint from "./SandboxedEditorHint";
import Spell from "../entities/items/Spell";

require("codemirror/mode/javascript/javascript");
require("codemirror/addon/hint/show-hint");
require("codemirror/lib/codemirror.css");
require("codemirror/addon/hint/show-hint.css");

export default class SandboxedEditorComponent extends Component {
  constructor(props) {
    super(props);
    this.state = {
      code: this.props.code
    };
  }

  componentDidMount() {
    // Initialize code hints. Must do it on next cycle so refs are resolved
    // and available.
    setTimeout(() => this.initCodeHints());
  }

  getCode() {
    return this.state.code;
  }

  initCodeHints() {
    const hint = new SandboxedEditorHint(
      this.codemirror.getCodeMirrorInstance()
    );
    const spell = new Spell("hints", this.props.api, "");
    const interpreter = spell.createInterpreter();
    hint.init(this.props.api, interpreter);
  }

  onClick = e => {
    if (this.state.isLoading) {
      GameEvent.absorbClick(e);
    }
  };

  onExecute = async e => {
    GameEvent.absorbClick(e);
    this.setState({ isLoading: true });
    try {
      const spell = new Spell("edit", this.props.api, this.state.code);
      await spell.cast();
      this.setState({ isLoading: false });
      GameEvent.fire(GameEvent.EDITOR_SUCCESS);
    } catch (e) {
      this.setState({ isLoading: false });
      GameEvent.fire(GameEvent.EDITOR_FAILURE);
    }
  };

  onUpdateCode = code => {
    this.setState({ code });
  };

  render() {
    const config = {
      extraKeys: { "Alt-Space": "autocomplete" },
      lineWrapping: true,
      mode: "javascript",
      tabSize: 2
    };

    const loader = this.state.isLoading ? <Indicators.Loader /> : "";
    const button = !this.state.isLoading ? (
      <div className={styles.executeButton}>
        <Button onClick={this.onExecute} text="Execute Incantation" />
      </div>
    ) : (
      ""
    );

    return (
      <div
        className={cx(styles.container, this.state.isLoading ? "loading" : "")}
        onClick={this.onClick}
      >
        {loader}
        <CodeMirror
          value={this.state.code}
          onChange={this.onUpdateCode}
          options={config}
          ref={codemirror => {
            this.codemirror = codemirror;
          }}
        />
        {button}
      </div>
    );
  }
}
