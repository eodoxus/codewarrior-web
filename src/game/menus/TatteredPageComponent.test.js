import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import GameEvent from "../engine/GameEvent";
import TatteredPageComponent from "./TatteredPageComponent";
import SandboxedEditorComponent from "./SandboxedEditorComponent";
import Spell from "../entities/items/Spell";

jest.mock("../engine/Audio");

SandboxedEditorComponent.prototype.render = jest.fn();

const spell = new Spell({}, "");

describe("<TatteredPageComponent />", () => {
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<TatteredPageComponent />, div);
    div.remove();
  });

  describe("Snapshots", () => {
    let node;
    beforeEach(() => {
      GameEvent.absorbClick = jest.fn();
      node = renderer.create(<TatteredPageComponent />);
      const instance = node.getInstance();
    });

    afterEach(() => {
      GameEvent.absorbClick.mockClear();
      GameEvent.removeAllListeners();
      node.unmount();
    });

    it("matches expected empty snapshot initially", () => {
      const tree = node.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("renders tattered page graphics container on open event", () => {
      GameEvent.fire(GameEvent.OPEN_TATTERED_PAGE, spell);
      const tree = node.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("renders empty on close event when opened", () => {
      GameEvent.fire(GameEvent.OPEN_TATTERED_PAGE, spell);
      GameEvent.fire(GameEvent.CLOSE_TATTERED_PAGE);
      const tree = node.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("closes when close button is clicked", () => {
      GameEvent.fire(GameEvent.OPEN_TATTERED_PAGE, spell);
      const close = node.root.findAllByType("div")[1];
      close.props.onClick({});
      const tree = node.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
