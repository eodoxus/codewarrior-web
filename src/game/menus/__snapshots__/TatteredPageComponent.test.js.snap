// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TatteredPageComponent /> Snapshots closes when close button is clicked 1`] = `""`;

exports[`<TatteredPageComponent /> Snapshots matches expected empty snapshot initially 1`] = `""`;

exports[`<TatteredPageComponent /> Snapshots renders empty on close event when opened 1`] = `""`;

exports[`<TatteredPageComponent /> Snapshots renders tattered page graphics container on open event 1`] = `
<div
  className="tattered-page"
  onClick={[Function]}
  style={
    Object {
      "backgroundImage": "url(/images/tattered-page.png)",
    }
  }
>
  <div
    className="undefined close"
    onClick={[Function]}
  >
    <span>
      x
    </span>
  </div>
  <div
    className={undefined}
  >
    <div
      className={undefined}
      onClick={[Function]}
    >
      <span>
        1
      </span>
      <div
        className=""
      />
    </div>
  </div>
  <div
    className={undefined}
  />
</div>
`;
